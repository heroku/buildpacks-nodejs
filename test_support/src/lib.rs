use libcnb_test::{
    assert_contains, BuildConfig, BuildpackReference, ContainerConfig, TestContext, TestRunner,
};
use std::fmt;
use std::fmt::Formatter;
use std::time::Duration;

const PORT: u16 = 8080;
const TIMEOUT: u64 = 10;

pub enum Builder {
    Heroku20,
    Heroku22,
}

impl fmt::Display for Builder {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        match *self {
            Builder::Heroku20 => write!(f, "heroku/buildpacks:20"),
            Builder::Heroku22 => write!(f, "heroku/builder:22"),
        }
    }
}

pub fn get_function_invoker_buildpacks() -> Vec<BuildpackReference> {
    vec![
        BuildpackReference::Other(String::from("heroku/nodejs-engine")),
        BuildpackReference::Other(String::from("heroku/nodejs-npm")),
        BuildpackReference::Crate,
    ]
}

pub fn get_yarn_buildpacks() -> Vec<BuildpackReference> {
    vec![
        BuildpackReference::Other(String::from("heroku/nodejs-engine")),
        BuildpackReference::Crate,
    ]
}

pub fn get_function_invoker_build_config(fixture: &str, builder: Builder) -> BuildConfig {
    BuildConfig::new(
        builder.to_string(),
        format!("../../test/fixtures/{fixture}"),
    )
    .buildpacks(get_function_invoker_buildpacks())
    .to_owned()
}

pub fn get_yarn_build_config(fixture: &str, builder: Builder) -> BuildConfig {
    BuildConfig::new(
        builder.to_string(),
        format!("../../test/fixtures/{fixture}"),
    )
    .buildpacks(get_yarn_buildpacks())
    .to_owned()
}

pub fn test_node_function(fixture: &str, builder: Builder, test_body: fn(TestContext)) {
    TestRunner::default().build(get_function_invoker_build_config(fixture, builder), |ctx| {
        test_body(ctx)
    });
}

pub fn test_yarn_app(fixture: &str, builder: Builder, test_body: fn(TestContext)) {
    TestRunner::default().build(get_yarn_build_config(fixture, builder), |ctx| {
        test_body(ctx)
    });
}

pub fn assert_health_check_responds(ctx: &TestContext) {
    ctx.start_container(ContainerConfig::new().expose_port(PORT), |container| {
        std::thread::sleep(Duration::from_secs(TIMEOUT));

        let addr = container
            .address_for_port(PORT)
            .expect("couldn't get container address");

        let resp = ureq::post(&format!("http://{0}", addr))
            .set("x-health-check", "true")
            .call()
            .expect("request to container failed")
            .into_string()
            .expect("response read error");

        assert_contains!(resp, "OK")
    })
}

pub fn assert_web_response(ctx: &TestContext, text: &'static str) {
    ctx.start_container(ContainerConfig::new().expose_port(PORT), |container| {
        std::thread::sleep(Duration::from_secs(5));

        let addr = container
            .address_for_port(PORT)
            .expect("couldn't get container address");

        let resp = ureq::get(&format!("http://{}/", addr))
            .call()
            .expect("request to container failed")
            .into_string()
            .expect("response read error");

        assert_contains!(resp, text);
    })
}
