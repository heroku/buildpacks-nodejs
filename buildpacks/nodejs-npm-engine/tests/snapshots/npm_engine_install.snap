---
source: test_support/src/lib.rs
---
## Heroku Node.js Engine

- Checking Node.js version
  - Detected Node.js version range: `16.20.2`
  - Resolved Node.js version: `16.20.2`
- Installing Node.js distribution
  - GET https://nodejs.org/download/release/v16.20.2/node-v16.20.2-<arch>.tar.gz ... (OK)
  - Downloading ... (<time_elapsed>)
  - Verifying checksum
  - Extracting Node.js `16.20.2 (<arch>)`
  - Installing Node.js `16.20.2 (<arch>)` ... (<time_elapsed>)
- Done (finished in <time_elapsed>)

## Heroku Node.js npm Engine

- Determining npm package information
  - Found `engines.npm` version `9.6.6` declared in `package.json`
  - GET https://registry.npmjs.org/npm ... (OK)
  - Downloading ... (<time_elapsed>)
  - Resolved npm version `9.6.6` to `9.6.6`
- Installing npm
  - GET https://registry.npmjs.org/npm/-/npm-9.6.6.tgz ... (OK)
  - Downloading ... (<time_elapsed>)
  - Removing npm bundled with Node.js
  - Installing requested npm
  - Successfully installed `npm@9.6.6`
- Done (finished in <time_elapsed>)

## Heroku Node.js npm Install

- Installing node modules
  - Using npm version `9.6.6`
  - Creating npm cache
  - Configuring npm cache directory
  - Running `npm ci "--production=false"`

      npm WARN config production Use `--omit=dev` instead.
      npm WARN old lockfile 
      npm WARN old lockfile The package-lock.json file was created with an old version of npm,
      npm WARN old lockfile so supplemental metadata must be fetched from the registry.
      npm WARN old lockfile 
      npm WARN old lockfile This is a one-time fix-up, please be patient...
      npm WARN old lockfile 

      added <NUMBER> packages, and audited <NUMBER> packages in <time_elapsed>

      found 0 vulnerabilities

  - Done (<time_elapsed>)
- Running scripts
  - No build scripts found
- Configuring default processes
  - Skipping default web process (no start script defined)
- Done (finished in <time_elapsed>)
